components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/SubscriberRegistrationInterceptor.java

```diff
@@ -0,0 +1,101 @@
+ */
+
+package org.wso2.carbon.apimgt.rest.api.util.interceptors;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.cxf.message.Message;
+import org.apache.cxf.phase.AbstractPhaseInterceptor;
+import org.apache.cxf.phase.Phase;
+import org.wso2.carbon.apimgt.api.APIConsumer;
+import org.wso2.carbon.apimgt.api.APIManagementException;
+import org.wso2.carbon.apimgt.api.model.Subscriber;
+import org.wso2.carbon.apimgt.impl.APIConstants;
+import org.wso2.carbon.apimgt.impl.utils.APIUtil;
+import org.wso2.carbon.apimgt.rest.api.util.utils.RestApiUtil;
+import org.wso2.carbon.context.PrivilegedCarbonContext;
+import org.wso2.carbon.registry.core.exceptions.RegistryException;
+
+public class SubscriberRegistrationInterceptor extends AbstractPhaseInterceptor {
+
+    private static final Log logger = LogFactory.getLog(SubscriberRegistrationInterceptor.class);
+    private static final String SUPER_TENANT_DOMAIN_NAME = "carbon.super";
+
+    public SubscriberRegistrationInterceptor() {
+        //We will use PRE_INVOKE phase as we need to process message before hit actual service
+        super(Phase.PRE_INVOKE);
+    }
+
+    /**
+     * Handles the incoming message after post authentication. Only used in Store REST API, to register a newly 
+     * signed up store user who hasn't logged in to Store for the first time either via REST API or Store UI. 
+     * This method will register the user as a subscriber 
+     * (register in AM_SUBSCRIBER table, add the default application for subscriber etc.).
+     *
+     * @param message cxf message
+     */
+    @Override
+    public void handleMessage(Message message) {
+        String username = RestApiUtil.getLoggedInUsername();
+        String groupId = RestApiUtil.getLoggedInUserGroupId();
+        String tenantDomain = RestApiUtil.getLoggedInUserTenantDomain();
+        try {
+            //takes a consumer object using the user set in thread local carbon context
+            APIConsumer apiConsumer = RestApiUtil.getLoggedInUserConsumer();
+            Subscriber subscriber = apiConsumer.getSubscriber(username);
+            if (subscriber == null) {
+                try {
+                    APIUtil.checkPermission(username, APIConstants.Permissions.API_SUBSCRIBE);
+                } catch (APIManagementException e) {
+                    // When user does not have subscribe permission we will log it and continue flow.
+                    // This happens when user tries to access anonymous apis although he does not have subscribe 
+                    // permission. It should be allowed.
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("User " + username + " does not have subscribe permission", e);
+                    }
+                    return;
+                }
+                if (!SUPER_TENANT_DOMAIN_NAME.equalsIgnoreCase(tenantDomain)) {
+                    loadTenantRegistry();
```

Comment:
Do we handle the cross tenant scenarios by switching the tenant flow properly in some other place.Is loading the tenant registry would be enough

Reply:
Yes it is handled at the AuthenticationInterceptor level, either from WebAppAuthenticatorImpl (if OAuth2) or from BasicAuthInterceptor (if basic auth).